### PR TITLE
Print optimal number of maPCA components and plot optimization curves

### LINKED ISSUE #834: Make MA-PCA criteria curves more accessible to users
### Summary

The number of PCA components is defined using one of several methods. [A recent neurostars question](https://neurostars.org/t/tedana-mask-and-components/21117/10) and an issue I've been examining would have been easier to answer if the criteria curves used to calculate the number of components were more accessible. I propose outputting these curves either by default or when the --verbose option is used.

### Next Steps

- @eurunuela has already made these values slightly more accessible in this PR: https://github.com/ME-ICA/mapca/pull/43 
- Have that PR also return either the full PCA object or the three criteria curves so that tedana can access that information
- Add code [after this line](https://github.com/ME-ICA/tedana/blob/64fa9d0b96df949a8ef38d450c37ca528f40f283/tedana/decomposition/pca.py#L236) to save the numbers in the criteria curves for aic, kic, and mdl.

### PR DESCRIPTION
**Pull request by eurunuela**
Closes #834 .

Changes proposed in this pull request:

- [x] Gets the optimal number of components and optimization curves for all three criteria in maPCA.
- [x] Plots the optimization curves of all three criteria in maPCA.
- [x] Prints the optimal number of components given by all three criteria to the logs. 
- [x] Print and plot optimal number of components given by 90% and 95% of variance explained when maPCA returns enough components.

Edit: this draft PR will fail until [#47](https://github.com/ME-ICA/mapca/pull/47) and [#48](https://github.com/ME-ICA/mapca/pull/48) are merged in maPCA.

### CODE REVIEW

**Comment by eurunuela** [comment]
@handwerkerd  I don't know what you want the `PCA_cross_component_metrics` file to contain exactly, so I am willing to leave that to you. What do you think of the changes I've made so far in this PR?

**Comment by handwerkerd** [comment]
I was hoping to read this carefully today, but it's not going to happen. I'll try to give a closer read soon. When I was discussing, `PCA_cross_component_metrics` it was in reference to `cross_component_metrics` that I've added to the component selection class in #756 Here is one example where it is used:
https://github.com/ME-ICA/tedana/blob/f011ba2c3a51eafd0a4d6429bbbfa4686baebd9d/tedana/selection/selection_nodes.py#L647-L652
The basic idea is to have a dictionary for all values that are calculated based on metrics across components (i.e. the kappa and rho elbows). The complements the component table, where each metric has a value for each component. In the current code, these cross component values are either in the log and nowhere else or not saved at all. Since you're criterion thresholds are the same general concept, I figured you could create a similar dictionary with a similar naming style.

**Comment by eurunuela** [comment]
The three-echo test should pass once we merge [PR #50 in maPCA](https://github.com/ME-ICA/mapca/pull/50).

**Comment by eurunuela** [comment]
The test will pass once [PR #50](https://github.com/ME-ICA/mapca/pull/50) is merged in maPCA.

**Comment by handwerkerd** [comment]
I just tried to locally run the three-echo test with mapca after [#50](https://github.com/ME-ICA/mapca/pull/50) was merged. The test is still failing in io.py with
```
Object of type ndarray is not JSON serializable
  File "[/Users/handwerkerd/code/tedana_community/me-ica/tedana/tedana/io.py]()", line 253, in save_json
    json.dump(data, fo, indent=4, sort_keys=True)
  File "[/Users/handwerkerd/code/tedana_community/me-ica/tedana/tedana/io.py]()", line 201, in save_file
    self.save_json(data, name)
  File "[/Users/handwerkerd/code/tedana_community/me-ica/tedana/tedana/decomposition/pca.py]()", line 322, in tedpca
    io_generator.save_file(mapca_results, "PCA cross component metrics json")
  File "[/Users/handwerkerd/code/tedana_community/me-ica/tedana/tedana/workflows/tedana.py]()", line 638, in tedana_workflow
    dd, n_components = decomposition.tedpca(
  File "[/Users/handwerkerd/code/tedana_community/me-ica/tedana/tedana/tests/test_integration.py]()", line 181, in test_integration_three_echo (Current frame)
    tedana_cli.tedana_workflow(
```
The four-echo test is passing so this isn't a universal issue. I'm not sure I have time to dig into this more right now, but wanted to share.

**Comment by tsalo** [comment]
I think we need to automatically convert numpy arrays to lists in `save_json`.

File: tedana/decomposition/pca.py
Comment (jbteves) [inline tedana/decomposition/pca.py:245]: This looks like we're accessing private components of the object. Could we either link to some documentation about the returned object or elaborate on what we're accessing here?

File: tedana/reporting/static_figures.py
Comment (jbteves) [inline tedana/reporting/static_figures.py:293]: This function has a large amount of repetition. Could we break it into smaller functions, parametrized perhaps by something like the input data and the label only, and if matplotlib requires this, the figure itself?

**Comment by jbteves** [comment]
Hi Eneko, I left some comments but didn't formally review. Let me know if you'd like for me to fix the merge conflicts and some of the refactors I mentioned as possibilities for you. However I do think it's worth considering some more documentation about how we're extracting these values from the mapca object.

File: tedana/decomposition/pca.py
Comment (tsalo) [reply tedana/decomposition/pca.py:245]: The trailing underscore indicates that the attribute is estimated from the data (via `fit`), rather than that it's private. We adopted this convention from scikit-learn.

**Comment by jbteves** [comment]
Hm, I merged the changes but it seems like some expected attribute is missing. It looks like CircleCI is just caching the last environment, so that it's not checking the versions in `setup.cfg` to update them appropriately, or perhaps some other error I'm missing. Any thoughts @eurunuela @tsalo ?

**Comment by jbteves** [comment]
@eurunuela I think you need to cut a new release with the changes so that we can pin that release.

**Comment by eurunuela** [comment]
> @eurunuela I think you need to cut a new release with the changes so that we can pin that release.

Done!

**Review by jbteves — DISMISSED** [review]
LGTM, thanks @eurunuela !

**Review by handwerkerd — CHANGES_REQUESTED** [review]
Mostly LGTM.
The one thing I notices is that a file called `./figures/pca_variance_explained.png` It looks similar to the relevant file `./figures/pca_criteria.png` but it just contains the vertical dashed lines without the criteria curves. My guess is that this was useful while working on the code, but isn't useful as a saved output. If I'm right, just don't save that file?

**Comment by eurunuela** [comment]
> Mostly LGTM. The one thing I notices is that a file called `./figures/pca_variance_explained.png` It looks similar to the relevant file `./figures/pca_criteria.png` but it just contains the vertical dashed lines without the criteria curves. My guess is that this was useful while working on the code, but isn't useful as a saved output. If I'm right, just don't save that file?

I'm sorry @handwerkerd, that was a mistake on my end. Here's what the figure should look like:
pca_variance_explained

I'm committing the fix now.

**Comment by handwerkerd** [comment]
Yay! That does look useful and I'm glad I noticed. In your sample figure above, I don't see a blue dashed line for AIC.

**Comment by eurunuela** [comment]
> Yay! That does look useful and I'm glad I noticed. In your sample figure above, I don't see a blue dashed line for AIC.

That's because it's right behind the KIC line. The figure was generated with the 3 echo integration test by the way.

**Pull request merged**
